pkg/subroutines/creator.go

```diff
@@ -0,0 +1,97 @@
+package subroutines
+
+import (
+	"context"
+	"fmt"
+
+	openfgav1 "github.com/openfga/api/proto/openfga/v1"
+	"github.com/openmfp/golang-commons/controller/lifecycle"
+	"github.com/openmfp/golang-commons/errors"
+	"go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	v1alpha1 "github.com/openmfp/account-operator/api/v1alpha1"
+)
+
+type CreatorSubroutine struct {
+	client  openfgav1.OpenFGAServiceClient
+	storeID string
+}
+
+func NewCreatorSubroutine(fgaGrpcAddr, storeId string) *CreatorSubroutine {
+	conn, _ := grpc.NewClient(fgaGrpcAddr,
+		grpc.EmptyDialOption{},
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithStatsHandler(otelgrpc.NewClientHandler()),
+	)
+
+	cl := openfgav1.NewOpenFGAServiceClient(conn)
+
+	return &CreatorSubroutine{client: cl, storeID: storeId}
+}
+
+func (e *CreatorSubroutine) Process(ctx context.Context, runtimeObj lifecycle.RuntimeObject) (ctrl.Result, errors.OperatorError) {
+	account := runtimeObj.(*v1alpha1.Account)
+
+	for _, condition := range account.Status.Conditions {
+		if condition.Reason == "AccountReady" && condition.Type == "Ready" && condition.Status == "True" {
+			return ctrl.Result{}, nil
+		}
+	}
+
+	_, err := e.client.Write(ctx, &openfgav1.WriteRequest{
+		StoreId: e.storeID,
+		Writes: &openfgav1.WriteRequestWrites{
+			TupleKeys: []*openfgav1.TupleKey{
+				{
+					Object:   fmt.Sprintf("account:%s", account.Name),
+					Relation: "owner",
+					User:     fmt.Sprintf("user:%s", *account.Spec.Creator),
+				},
+			},
+		},
+	})
+
+	if err != nil {
```

Comment:
additionally for some reason it could be the case, that the tuple is already part of the OpenFGA graph. This will then lead to an error on this function call as you are trying to write a duplicated tuple. This would lead to the subroutine and the resource being in an error state. But actually, if the tuple is already present we have already achieved the goal.

Therefore we have a helper in our shared library: https://github.com/openmfp/golang-commons/blob/main/fga/helpers/helper.go#L70

Which makes sure to check for such an error, so that we can react to it properly.

The usage would look like:

```go
if helpers.IsDuplicateWriteError(err) {
  err = nil
}

if err != nil {
 // ...
}

Reply:
int32(s.Code()) == int32(openfgav1.ErrorCode_write_failed_due_to_invalid_input)

It doesn't check for duplicates, but for general write error. Am I right?

Reply:
The invalid input error is the one that is retrieved on a duplicate write, so its the right thing to check